From 50d736664f5140bae008644a6fc0fee42e54c13a Mon Sep 17 00:00:00 2001
From: Paul Holzinger <pholzing@redhat.com>
Date: Thu, 14 Aug 2025 14:07:37 +0200
Subject: [PATCH 1/3] test: replace ncat with socat

ncat keeps causing issues with different versions having different
behaviors. And many distro ship different incompatible versions so it is
hard to make the test work everywhere consistently.

To address replace ncat with socat. Socat seems to much more stable and
we have been using it in podman for pasta tests for a while without
problems.

Fixes: #622

Signed-off-by: Paul Holzinger <pholzing@redhat.com>
---
 test/100-basic-name-resolution.bats |  7 ++++---
 test/600-errors.bats                | 16 ++++++++--------
 2 files changed, 12 insertions(+), 11 deletions(-)

diff --git a/test/100-basic-name-resolution.bats b/test/100-basic-name-resolution.bats
index 40fbd1e5..e81639c0 100644
--- a/test/100-basic-name-resolution.bats
+++ b/test/100-basic-name-resolution.bats
@@ -74,8 +74,8 @@ function teardown() {
 @test "basic container - dns itself (bad and good should fall back)" {
 	setup_dnsmasq
 
-	# using sh-exec to keep the udp query hanging for at least 3 seconds
-	nsenter -m -n -t $HOST_NS_PID ncat -l -u 127.5.5.5 53 --sh-exec "sleep 3" 3>/dev/null &
+	# using exec to keep the udp query hanging for at least 3 seconds
+	nsenter -m -n -t $HOST_NS_PID socat UDP4-LISTEN:53,bind=127.5.5.5 EXEC:"sleep 3" 3>/dev/null &
 	HELPER_PID=$!
 
 	subnet_a=$(random_subnet 5)
@@ -90,8 +90,9 @@ function teardown() {
 	run_in_container_netns "$a1_pid" "dig" "$TEST_DOMAIN" "@$gw"
 	assert "$output" =~ "Query time: [23][0-9]{3} msec" "timeout should be 2.5s so request should then work shortly after (udp)"
 
+	kill -9 "$HELPER_PID" || true
 	# Now the same with tcp.
-	nsenter -m -n -t $HOST_NS_PID ncat -l 127.5.5.5 53 --sh-exec "sleep 3" 3>/dev/null &
+	nsenter -m -n -t $HOST_NS_PID socat TCP4-LISTEN:53,bind=127.5.5.5 EXEC:"sleep 3" 3>/dev/null &
 	HELPER_PID=$!
 	run_in_container_netns "$a1_pid" "dig" +tcp "$TEST_DOMAIN" "@$gw"
 	assert "$output" =~ "Query time: [23][0-9]{3} msec" "timeout should be 2.5s so request should then work shortly after (tcp)"
diff --git a/test/600-errors.bats b/test/600-errors.bats
index 0b1b6844..273a13ff 100644
--- a/test/600-errors.bats
+++ b/test/600-errors.bats
@@ -6,10 +6,10 @@
 load helpers
 
 
-NCPID=
+SOCAT_PID=
 
 function teardown() {
-    kill -9 $NCPID
+    kill -9 $SOCAT_PID
     basic_teardown
 }
 
@@ -17,10 +17,10 @@ function teardown() {
 @test "aardvark-dns should fail when udp port is already bound" {
 	# bind the port to force a failure for aardvark-dns
 	# we cannot use run_is_host_netns to run in the background
-	nsenter -m -n -t $HOST_NS_PID ncat -u -l 0.0.0.0 53 </dev/null 3> /dev/null &
-	NCPID=$!
+	nsenter -m -n -t $HOST_NS_PID socat UDP4-LISTEN:53 - 3> /dev/null &
+	SOCAT_PID=$!
 
-	# ensure nc has time to bind the port
+	# ensure socat has time to bind the port
 	sleep 1
 
 	subnet_a=$(random_subnet 5)
@@ -33,10 +33,10 @@ function teardown() {
 @test "aardvark-dns should fail when tcp port is already bound" {
 	# bind the port to force a failure for aardvark-dns
 	# we cannot use run_is_host_netns to run in the background
-	nsenter -m -n -t $HOST_NS_PID ncat -l 0.0.0.0 53 </dev/null 3> /dev/null &
-	NCPID=$!
+	nsenter -m -n -t $HOST_NS_PID socat TCP4-LISTEN:53 - 3> /dev/null &
+	SOCAT_PID=$!
 
-	# ensure nc has time to bind the port
+	# ensure socat has time to bind the port
 	sleep 1
 
 	subnet_a=$(random_subnet 5)

From 071143f5ea92ecd50d198e45c3919b8f6eb982d4 Mon Sep 17 00:00:00 2001
From: Paul Holzinger <pholzing@redhat.com>
Date: Thu, 14 Aug 2025 14:12:21 +0200
Subject: [PATCH 2/3] rpm: tests require socat instead of ncat now

Signed-off-by: Paul Holzinger <pholzing@redhat.com>
---
 rpm/aardvark-dns.spec | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/rpm/aardvark-dns.spec b/rpm/aardvark-dns.spec
index 13c34633..6a1866e0 100644
--- a/rpm/aardvark-dns.spec
+++ b/rpm/aardvark-dns.spec
@@ -61,7 +61,7 @@ Requires: bats
 Requires: bind-utils
 Requires: jq
 Requires: netavark
-Requires: nmap-ncat
+Requires: socat
 Requires: dnsmasq
 
 %description tests
@@ -97,7 +97,7 @@ tar fx %{SOURCE1}
 
 %{__install} -d -p %{buildroot}%{_datadir}/%{name}/test
 %{__cp} -rp test/* %{buildroot}%{_datadir}/%{name}/test/
-%{__rm} -rf %{buildroot}%{_datadir}/%{name}/test/tmt/    
+%{__rm} -rf %{buildroot}%{_datadir}/%{name}/test/tmt/
 
 # Add empty check section to silence rpmlint warning.
 # No tests meant to be run here.

From 32deed993c33b2481a39c423f67842e60f181a3c Mon Sep 17 00:00:00 2001
From: Paul Holzinger <pholzing@redhat.com>
Date: Wed, 20 Aug 2025 17:07:11 +0200
Subject: [PATCH 3/3] update CI images

Build in https://github.com/containers/automation_images/pull/415.

Signed-off-by: Paul Holzinger <pholzing@redhat.com>
---
 .cirrus.yml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/.cirrus.yml b/.cirrus.yml
index 0b31ac75..efd8fb4e 100644
--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -19,7 +19,7 @@ env:
     NETAVARK_URL: "https://api.cirrus-ci.com/v1/artifact/github/containers/netavark/success/binary.zip?branch=${NETAVARK_BRANCH}"
     # Save a little typing (path relative to $CIRRUS_WORKING_DIR)
     SCRIPT_BASE: "./contrib/cirrus"
-    IMAGE_SUFFIX: "c20250422t130822z-f42f41d13"
+    IMAGE_SUFFIX: "c20250820t132717z-f42f41d13"
     FEDORA_NETAVARK_IMAGE: "fedora-netavark-${IMAGE_SUFFIX}"
     FEDORA_NETAVARK_AMI: "fedora-netavark-aws-arm64-${IMAGE_SUFFIX}"
     EC2_INST_TYPE: "t4g.xlarge"
